nl.nikhef.pdp.fetchcrl/tags/fetch-crl-3.0.11/clean-crl.8
.\" "@(#)$Id$"
.\"
.\" 
.TH CLEAN-CRL 8 local "Trust Anchor Utilities"
.SH NAME
clean-crl \- remove orphaned CRL like files from a certificate directory
.SH SYNOPSIS
.ll +8
.B clean-crl
.RB [ \-l\ crlpath ]
.RB [ \-v ]
.RB [ \-V ]
.RB [ \-n ]
.RB [ \-h ]
.ll -8
.SH DESCRIPTION
The 
.I clean-crl
utility will remove CRL like files named
.IR hash .r n
from the directory specified with the 
.B \-l
option if there is no corresponding 
.RI . n
file in the same.
In effect, if the directory is solely used to hold CA certificates
in the common OpenSSL format, it will thus remove CRL files for
which the corresponding CA does  not or no longer exists in the
directory.

.SH OPTIONS
.TP
.B \-h --help
Show help text.
.TP
.B \-l --cadir metadata-directory
The script will search this directory for files with the
suffix 
.RI .r i .
There is no default - a common choice is /etc/pki/tls/certs, 
/etc/openldap/cacerts, or /etc/grid-security/certificates.

.TP
.B \-V --version 
Display version number (same as corresponding fetch-crl)

.TP
.B \-v --verbose
Verbose mode

.TP
.B \-n --dryrun
Do not actually remove any files (useful primarily with -v)

.SH CONFIGURATION
None.

.SH NOTES
This tool does not check the contents of the files removed, and will
blindly unlink any file which even removely looks like an OpenSSL CRL
file. Use with extreme caution.

.SH "SEE ALSO"
fetch-crl(8), openssl(1), 
http://wiki.nikhef.nl/grid/FetchCRL3

.SH "DIAGNOSTICS"
Exit status is normally 0;
if an error occurs, exit status is 1 and diagnostics will be written
to standard error.

.SH LICENSE
Licensed under the Apache License, Version 2.0 (the "License"); 

.B http://www.apache.org/licenses/LICENSE-2.0

.SH BUGS
Does not check the contents of the files removed. 



CHANGES
==============================================================================
CHANGES to fetch-crl - the Certificate Revocation List retrieval tool
==============================================================================
The fetch-crl utility will retrieve certificate revocation lists (CRLs) for
a set of installed trust anchors, based on crl_url files or IGTF-style info
files. It will install these for use with OpenSSL, NSS or third-party tools.

Changes in 3.0.11-1
----------------------
* Added reference to /etc/fetch-crl.d/ to the man page, used shortened URL
  to full documentation in man page

Changes in 3.0.10-1
----------------------
* Added a "noquiet" option in the configuration file that will override
  the default single "-q" option in the cro-job that is shipped with
  the fetch-crl3 init scripts (feature request by Ryan Taylor)
* Added option "--inet6glue" and "inet6glue" config setting to load 
  the Net::INET6Glue perl module (if it is available) to use IPv6 
  connections in LWP to download CRLs

Changes in 3.0.8-1
----------------------
* Trust anchor name inferrence based on retrieved-CRL added as option (at cost 
  of retrieving CRL even if there is no accompanying trust anchor found later)
  Option is disabled by default, but can be enabled by using @HASH@ in the 
  ca-template name list. (feature request by Rob van der Wal, SARA, NL)

Changes in 3.0.7-1
----------------------
* CRL modofication time heuristic inadvertently modified file name templates
  (solves issue kindly reported by Elan Ruusamae)
* Expanded representation of tokenisation characters in strings to work 
  around bug in file(1) (rhbz#699546, works around RedHat Bugzilla 699548)

Changes in 3.0.6-1
----------------------
* Response parsing disabled to suppress superfluous warning on unexpected 
  UTF-8 respons when retrieving a CRL (solves RedHat Bugzilla 688902)

Changes in 3.0.5-1
----------------------
* CRLs for multiple similarly-named trust anchors might not all be downloaded.
  This is fixed in this release.
* Spurious "restoreLogMode" internal errors are no longer raised

Changes in 3.0.4-1
----------------------
* Add support for directory based drop-in configuration in /etc/fetch-crl.d/
* Only use cached CRL contents if the nextUpdate time of the cached CRL is 
  still in the future. This will ensure that a new download is attempted 
  each and everytime for CRLs that have already expired.

Changes in 3.0.3-1
----------------------
* Clean up of man page format macro PU (reported by Mattias Ellert)

Changes in 3.0.2-1
----------------------
* Clean up of man page format macro PU (reported by Mattias Ellert)

Changes in 3.0.1-1
----------------------
* hunts through more places to find the latest successful CRL download to
  set the latest local modification time for a CRL 
  (resolves a comparison error in case output and infodir are unset)

Changes in 3.0.0-0.RC4
----------------------
* the config file name has changed to fetch-crl.conf, although a 
  fetch-crl.cnf file will also be used when present
* symlinked meta-data files can be ignored with the --nosymlinks option
  (or nosymlinks in the configuration file). This allows fetch-crl to be
  used effectively with new-format IGTF distribution before 1.37
* infinite loop for non-indexed CA file names fixed

Changes in fetch-crl 3.0
------------------------
* fetch-crl 3.0 is a complete re-write, and shares no code with the 1.x and
  2.x series utility of the same name, although the function and some of
  the syntax is obviously the same

* support for multiple output formats: OpenSSL 1 in dual-hash mode, specific
  DER and PEM outputs, and NSS databases
* support for multiple CRLs for a single CA, allowing more than one CA with
  the same subject name but different CLRs. Review your client software to see
  if and how these CRLs are used.
* stateful retrieval helps reduce bandwidth usage by caching the CRLs locally
  and respecting the Cache Control headers sent by the web server hosting the
  CRL. This can reduce the number of downloads
* support for HEAD-only requests when state preservation is used (initially
  only retrieve HTTP headers, and only if the CRL actually changed to a full
  download)
* support for more CRL retrieval protocols (file:// and ftp://)
* ability to try site-local URLs first, before relying on the URLs shipped with
  the trust anchor. This allows building an explicit local caching (web) server.
* ability to specify additional URLs to try in case the URLs shipped with the
  trust anchor were not responsive. This allows for automatic fall-back to
  (local or global) mirror services for CRL downloads
* warnings and errors can be suppressed on a per-trust anchor basis, to allow
  silencing for particularly unstable trust anchors
* aging tolerance (the delay time before errors are generated in case downloads
  consistently fail) can be configured on a per-trust anchor basis
* parallel downloading for multiple trust anchors
* minimized use of temporary files in the file system (now limited to the
  invocation of OpenSSL only, and only for brief periods of time)
* dependencies on wget, lynx and other unix utilities have been removed
* explicit web proxy support (using LWP http proxies)
* completely re-written in perl, with some (hopefully minimal) dependencies:
  LWP, Sys::Syslog, POSIX. And Data::Dumper (when debugging is enabled),
  and IO::Select (if parallel downloads are enabled). 

Differences with respect to the previous versions

* when downloading CRLs via https, the server certificate is not checked, 
  neither for the correct DNS name nor for being issued by a valid CA. Since 
  the CRL in itself is signed, this is not a security vulnerability. If 
  stricter checking is anyway desired, and the Crypt::SSLeay perl module has 
  been installed, set the HTTPS_CA_FILE environment variable before invoking 
  fetch-crl -- but keep in mind that the DNS name verification is limited 

nl.nikhef.pdp.fetchcrl/trunk/clean-crl.pl
#! /usr/bin/perl -w
#
use strict;
use Getopt::Long qw(:config no_ignore_case bundling);
use POSIX;

my $versionid = "20130702.01";

my $targetdir;
my $show_help;
my $show_version;
my $verbose;
my $dryrun;

sub help() {
  (my $name = $0) =~ s/.*\///;
  print <<EOHELP;
The $name utility will eradicate [0-9a-f]{8}.r\\d+ files from
the directory given to the "-l" option if no matching [0-9a-f]{8}.\\d+
file can be found in the same, which in most cases will wipe stale 
historic CRLs from an X509_CERT_DIR like directory.
Use at your own risk. It may be wiping files that you would have
liked to keep, or it may kill your pet.

Options:
  -l | --cadir <path>
    directory to cleanse of old CRL-ish files
  -v[v...] | --verbose 
    become more verbose and talkative
  -n | --dryrun
    do not actually unlink any files
  -V | --version
    show a version number
  -h | --help
    this help text

Examples:
  $name -l /etc/grid-security/certificates

Diagnostics:
  ". not found": consult an expert.

EOHELP
  return 1;
}

sub showversion() {
  (my $name = $0) =~ s/.*\///;
  print "$name version $versionid\n";
  return 1;
}

&GetOptions(
  "l|cadir=s" => \$targetdir,
  "n|dryrun" => \$dryrun,
  "h|help" => \$show_help,
  "v|verbose+" => \$verbose,
  "V|version" => \$show_version
  ) or &help and exit(1);

$show_help and &help() and exit (0);
$show_version and &showversion() and exit (0);
$verbose = 0 unless defined $verbose;
$dryrun = 0 unless defined $dryrun;

die "Error: target directory undefined, please supply -l argument!\n" 
  unless $targetdir;
die "Error: target directory $targetdir does not exist\n"
  unless -e $targetdir;
die "Error: target directory $targetdir is not a directory\n"
  unless -d $targetdir;

# read the directory and find all CA like .\d and CRL like files, 
# recoding the hashes of the info files in an array, and then in a
# second pass weeding out those CRL ".r*" files that do not have
# a corresponding info or crl_url file
# the remainer is a candidate for deletion
my $dh;
my @crlfiles;
my %infohashes;
opendir($dh,$targetdir) or die "Cannot open $targetdir: $!\n";
while ( my $fn = readdir $dh ) {
  $fn =~ /^([0-9a-f]{8})\.(\d+)$/ and do {
    $infohashes{$1}=1;
    ($verbose > 2) and print "Hash $1 belongs to an active CA\n";
  };
  $fn =~ /^([0-9a-f]{8})\.r(\d+)$/ and do {
    push @crlfiles,$fn;
    ($verbose > 2) and print "File $fn is classified as a CRL file\n";
  };
}

my @candidates = grep {
    /^([0-9a-f]{8})\.r([0-9]+)$/;
    ! exists $infohashes{$1};
  } @crlfiles;

$verbose > 0 and do {
  if ( $#candidates >= 0 ) {
    print "The following CRL like files are about to be deleted".
      ($dryrun?" ... NOT!":".")."\n";
    foreach my $fn ( @candidates ) { print "  $fn\n"; }
  } else {
    print "No orphaned CRL like files found in $targetdir\n";
  }
};

if ( ! $dryrun ) {
  foreach my $fn ( @candidates ) { 
    unlink("$targetdir/$fn") or warn "Cannot remove $targetdir/$fn: $!\n";
  }
}

1;
